File: apps/android/core/crypto/src/main/java/app/ollo/crypto/SessionVault.kt
package app.ollo.crypto

/**
 * Access / refresh live in a dedicated wrapped slot. They are never written
 * into backups. Wipe this on logout, refresh reuse, and remote wipe.
 */
data class SessionSecrets(
    val userId: String,
    val deviceId: String,
    val access: String,
    val refresh: String,
)

class SessionVault(
    private val store: IdentityStore,
    private val wrapKey: ByteArray,
) {
    fun save(secrets: SessionSecrets) {
        store.put(
            wrapKey,
            KEY,
            BlobMap.encode(
                linkedMapOf(
                    "userId" to secrets.userId.toByteArray(Charsets.UTF_8),
                    "deviceId" to secrets.deviceId.toByteArray(Charsets.UTF_8),
                    "access" to secrets.access.toByteArray(Charsets.UTF_8),
                    "refresh" to secrets.refresh.toByteArray(Charsets.UTF_8),
                ),
            ),
        )
    }

    fun load(): SessionSecrets? {
        val raw = store.get(wrapKey, KEY) ?: return null
        val map = BlobMap.decode(raw)
        val userId = map["userId"]?.toString(Charsets.UTF_8) ?: return null
        val deviceId = map["deviceId"]?.toString(Charsets.UTF_8) ?: return null
        val access = map["access"]?.toString(Charsets.UTF_8) ?: return null
        val refresh = map["refresh"]?.toString(Charsets.UTF_8) ?: return null
        return SessionSecrets(userId, deviceId, access, refresh)
    }

    fun clear() {
        store.remove(KEY)
    }

    companion object {
        const val KEY = "session.v1"
    }
}

File: apps/ios/Sources/OlloCrypto/SessionVault.swift
import Foundation

/// Access / refresh live in a dedicated wrapped slot. Never written into backups.
public struct SessionSecrets: Sendable, Equatable {
    public var userId: String
    public var deviceId: String
    public var access: String
    public var refresh: String

    public init(userId: String, deviceId: String, access: String, refresh: String) {
        self.userId = userId
        self.deviceId = deviceId
        self.access = access
        self.refresh = refresh
    }
}

public final class SessionVault: @unchecked Sendable {
    public static let key = "session.v1"
    private let store: IdentityStore
    private let wrapKey: Data

    public init(store: IdentityStore, wrapKey: Data) {
        self.store = store
        self.wrapKey = wrapKey
    }

    public func save(_ secrets: SessionSecrets) throws {
        try store.put(
            wrapKey: wrapKey,
            key: Self.key,
            plaintext: try BlobMap.encode([
                "userId": Data(secrets.userId.utf8),
                "deviceId": Data(secrets.deviceId.utf8),
                "access": Data(secrets.access.utf8),
                "refresh": Data(secrets.refresh.utf8),
            ])
        )
    }

    public func load() throws -> SessionSecrets? {
        guard let raw = try store.get(wrapKey: wrapKey, key: Self.key) else { return nil }
        let map = try BlobMap.decode(raw)
        guard
            let userId = string(map["userId"]),
            let deviceId = string(map["deviceId"]),
            let access = string(map["access"]),
            let refresh = string(map["refresh"])
        else { return nil }
        return SessionSecrets(userId: userId, deviceId: deviceId, access: access, refresh: refresh)
    }

    public func clear() throws {
        try store.remove(key: Self.key)
    }

    private func string(_ data: Data?) -> String? {
        guard let data else { return nil }
        return String(data: data, encoding: .utf8)
    }
}
